sdk/core/azure-core/inc/nullable.hpp

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <cstdlib> // for abort
+#include <new> // for placement new
+#include <type_traits>
+#include <utility> // for swap and move
+
+namespace Azure { namespace Core {
+
+  template <class T> class Nullable {
+    union
+    {
+      char m_disengaged; // due to constexpr rules for the default constructor
+      T m_value;
+    };
+
+    bool m_hasValue;
+
+    void Destroy() noexcept /* enforces termination */
+    {
+      if (m_hasValue)
+      {
+        m_value.~T();
+        m_hasValue = false;
+      }
+    }
+
+  public:
+    constexpr Nullable() : m_disengaged{}, m_hasValue(false) {}
+    constexpr Nullable(const T& initialValue) : m_value(initialValue), m_hasValue(true) {}
+
+    Nullable(const Nullable& other) noexcept(std::is_nothrow_copy_constructible<T>::value)
+        : m_hasValue(other.m_hasValue)
+    {
+      if (m_hasValue)
+      {
+        // standard implementation has to use std::addressof instead of & :)
+        ::new (static_cast<void*>(&m_value)) T(other.m_value);
+      }
+    }
+
+    Nullable(Nullable&& other) noexcept(std::is_nothrow_move_constructible<T>::value)
+        : m_hasValue(other.m_hasValue)
+    {
+      if (m_hasValue)
+      {
+        // standard implementation has to use std::addressof instead of & :)
+        ::new (static_cast<void*>(&m_value)) T(std::move(other.m_value));
+      }
+    }
+
+    ~Nullable()
+    {
+      if (m_hasValue)
+      {
+        m_value.~T();
+      }
+    }
+
+    void Reset() noexcept /* enforces termination */ { Destroy(); }
+
+    // this assumes that swap can't throw if T is nothrow move constructible because
+    // is_nothrow_swappable is added in C++17
+    void Swap(Nullable& other) noexcept(std::is_nothrow_move_constructible<T>::value)
+    {
+      if (m_hasValue)
+      {
+        if (other.m_hasValue)
+        {
+          using std::swap;
+          swap(m_value, other.m_value);
+        }
+        else
+        {
+          ::new (static_cast<void*>(&other.m_value)) T(std::move(m_value)); // throws
+          other.m_hasValue = true;
+          Destroy();
+        }
+      }
+      else if (other.m_hasValue)
+      {
+        ::new (static_cast<void*>(&m_value)) T(std::move(other.m_value)); // throws
+        m_hasValue = true;
+        other.Destroy();
+      }
+    }
+
+    // Intentionally lowercase to override the swap
```

Comment:
```suggestion
    // Intentionally lowercase to follow the Swappable requirements
```

https://en.cppreference.com/w/cpp/named_req/Swappable